tutorials/embedded-c-getting-started/port/src/iotc_bsp_io_net_posix.c

```diff
@@ -0,0 +1,288 @@
+  const int flags = fcntl(*iotc_socket, F_GETFL, 0);
+
+  if (-1 == flags || -1 == fcntl(*iotc_socket, F_SETFL, flags | O_NONBLOCK)) {
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+
+  return IOTC_BSP_IO_NET_STATE_OK;
+}
+
+iotc_bsp_io_net_state_t iotc_bsp_io_net_connect(iotc_bsp_socket_t* iotc_socket,
+                                                const char* host,
+                                                uint16_t port) {
+  struct hostent* hostinfo = gethostbyname(host);
+
+  /* if null it means that the address has not been found */
+  if (NULL == hostinfo) {
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+
+  struct sockaddr_in name = {
+      .sin_family = AF_INET,
+      .sin_port = htons(port),
+      .sin_addr = *((struct in_addr*)hostinfo->h_addr_list[0]),
+      .sin_zero = {0}};
+
+  if (-1 ==
+      connect(*iotc_socket, (struct sockaddr*)&name, sizeof(struct sockaddr))) {
+    return (EINPROGRESS == errno) ? IOTC_BSP_IO_NET_STATE_OK
+                                  : IOTC_BSP_IO_NET_STATE_ERROR;
+  } else {
+    // todo_atigyi: what to do here?
+    // does this mean the socket is BLOCKING?
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+
+  return IOTC_BSP_IO_NET_STATE_ERROR;
+}
+
+iotc_bsp_io_net_state_t iotc_bsp_io_net_connection_check(
+    iotc_bsp_socket_t iotc_socket, const char* host, uint16_t port) {
+  (void)host;
+  (void)port;
+
+  int valopt = 0;
+  socklen_t lon = sizeof(int);
+
+  if (getsockopt(iotc_socket, SOL_SOCKET, SO_ERROR, (void*)(&valopt), &lon) <
+      0) {
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+
+  if (valopt) {
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+
+  return IOTC_BSP_IO_NET_STATE_OK;
+}
+
+iotc_bsp_io_net_state_t iotc_bsp_io_net_write(iotc_bsp_socket_t iotc_socket,
+                                              int* out_written_count,
+                                              const uint8_t* buf,
+                                              size_t count) {
+  if (NULL == out_written_count || NULL == buf) {
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+
+  int errval = 0;
+  socklen_t lon = sizeof(int);
+
+  if (getsockopt(iotc_socket, SOL_SOCKET, SO_ERROR, (void*)(&errval), &lon) <
+      0) {
+    errval = errno;
+    errno = 0;
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+
+  if (errval != 0) {
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+
+  *out_written_count = write(iotc_socket, buf, count);
+
+  if (*out_written_count < 0) {
+    *out_written_count = 0;
+
+    errval = errno;
+    errno = 0;
+
+    if (EAGAIN == errval) {
+      return IOTC_BSP_IO_NET_STATE_BUSY;
+    }
+
+    if (ECONNRESET == errval || EPIPE == errval) {
+      return IOTC_BSP_IO_NET_STATE_CONNECTION_RESET;
+    }
+
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+
+  return IOTC_BSP_IO_NET_STATE_OK;
+}
+
+iotc_bsp_io_net_state_t iotc_bsp_io_net_read(iotc_bsp_socket_t iotc_socket,
+                                             int* out_read_count, uint8_t* buf,
```

Comment:
Prefer to avoid the hanging indent, just:

```c
iotc_bsp_io_net_state_t iotc_bsp_io_net_read(
        iotc_bsp_socket_t iotc_socket, int* out_read_count, uint8_t* buf,
        size_t count) {
```